File: 백준/Silver/2583. 영역 구하기/영역 구하기.cc
#include <bits/stdc++.h>

using namespace std;

const int dy[] = { -1, 0, 1, 0 };
const int dx[] = { 0, 1, 0 ,-1 };
int M, N, K, ret, xa, ya, xb, yb;
vector<int> v;
int a[102][102];
int visited[102][102];

void DFS(int y, int x) {
	visited[y][x] = 1;
	ret++;
	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i];
		int nx = x + dx[i];
		if (ny < 0 || nx < 0 || ny >= M || nx >= N) continue;
		if (!visited[ny][nx] && a[ny][nx] == 1) {
			DFS(ny, nx);
		}
	}
	return;
}
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);

	cin >> M >> N >> K;

	fill(&a[0][0], &a[0][0] + 102 * 102, 1);

	for (int i = 0; i < K; i++) {
		cin >> xa >> ya >> xb >> yb;

		for (int i = ya; i < yb; i++) {
			for (int j = xa; j < xb; j++) {
				a[i][j] = 0;
				visited[i][j] = 1;
			}
		}
	}

	for (int i = 0; i < M; i++) {
		for (int j = 0; j < N; j++) {
			if (!visited[i][j] && a[i][j] == 1) {
				ret = 0;
				DFS(i, j);
				v.push_back(ret);
			}
		}
	}
	sort(v.begin(), v.end());
	cout << v.size() << '\n';
	for (int i = 0; i < v.size(); i++) {
		cout << v[i] << " ";
	}

	return 0;
}

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static int[][] maps;
    static int[][] visited;
    static int[] dy = {-1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};
    static int area;

    static void dfs(int y, int x) {
        visited[y][x] = 1;
        area++;

        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny < 0 || nx < 0 || ny >= maps.length || nx >= maps[0].length || visited[ny][nx] == 1) {
                continue;
            }
            if (maps[ny][nx] == 0) {
                dfs(ny, nx);
            }
        }
    }


    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(bf.readLine());

        int M = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        maps = new int[N][M];
        visited = new int[N][M];

        for (int i = 0; i < K; i++) {
            StringTokenizer line = new StringTokenizer(bf.readLine());
            int startY = Integer.parseInt(line.nextToken());
            int startX = Integer.parseInt(line.nextToken());
            int endY = Integer.parseInt(line.nextToken());
            int endX = Integer.parseInt(line.nextToken());
            for (int j = startY; j < endY; j++) {
                for (int k = startX; k < endX; k++) {
                    maps[j][k] = 1;
                }
            }
        }
        List<Integer> list = new ArrayList<>();
        int ret = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (visited[i][j] == 0 && maps[i][j] == 0) {
                    area = 0;
                    dfs(i, j);
                    list.add(area);
                    ret++;
                }
            }
        }
        System.out.println(ret);
        Collections.sort(list);
        for (int i = 0; i < list.size(); i++) {
            System.out.print(list.get(i) + " ");
        }
    }
}